### ISSUE TITLE
AArch64: Improve code generation for array access

### ISSUE THREAD

**Issue by knn-k** [eclipse/omr#6087]
AArch64 array access code requires extra instruction and extra register allocation.

Example (SIMD array + array):
```
	sbfmx	w2, w15, 62, 31
	addx	w0, w12, w2
	addx	w1, w13, w2
	vldrimmq	s2, [w1, 0]
	addx	w1, w12, w2
	vldrimmq	s1, [w1, 0]
	vadd4s	s0, s2, s1
	vstrimmq	s0, [w0, 0]
```

This sequence (8 instructions) can be changed to the following 5 instructions.  You don't need to allocate `w0` and `w1` in this sequence:
```
	sbfmx 	w2, w15, 62, 31
	vldroffq 	s2, [w13, w2]
	vldroffq 	s1, [w12, w2]
	vadd4s 	s0, s2, s1
	vstroffq 	s0, [w12, w2]
```

**Comment by knn-k** [comment]
Related PR: #5887 (constant offset case)

**Comment by knn-k** [comment]
An idea to implement this is to introduce abstract load/store opcodes that don't have addressing mode information.
`populateMemoryReference()` will set `baseReg`, `immOffset`, or `idxReg`, then `generateBinaryEncoding()` will choose the appropriate load/store instruction based on the addressing mode.

**Comment by knn-k** [comment]
1. Introduce new load/store opcodes that have type information but addressing mode: `loadb`, `loadh`, `loadw`, `loadx`, `loads`, `loadd`, `loadq`, for example.
2. Replace the load/store opcodes used in load/store evaluators, such as `iloadEvaluator()` by the new opcodes above.  Don't replace the load/store opcodes used in other places.  You can use load/store opcodes with explicit addressing mode where you want.
3. Improve `populateMemoryReference()`: Set immediate offset or index register appropriately in memory references for array access.  It needs to take care of the cases where the immediate offset does not fit in the final instruction.  In such cases, you use the `[baseReg, idxReg]` form by generating `mov idxReg, #imm`.
4. Improve `generateBinaryEncoding()`: When the function finds the load/store opcodes introduced in 1, it chooses the instruction based on the addressing mode.  For example, `loadw` can be converted to one of `ldrimmw`, `ldurw`, `ldroffw`.

**Comment by Akira1Saitoh** [comment]
I have found that the reference count of `aladd`  node is not decremented in constant offset case:
```
------------------------------
 n15n     (  0)  ireturn                                                                              [    0xffff668d3460] bci=[-1,3,19] rc=0 vc=139 vn=- li=2 udi=- nc=1
 n14n     (  1)    iloadi  <array-shadow>[#204  Shadow] [flags 0x80000603 0x0 ] (cannotOverflow )     [    0xffff668d3410] bci=[-1,2,19] rc=1 vc=139 vn=- li=2 udi=- nc=1 flg=0x1000
 n13n     (  1)      aladd (X>=0 internalPtr sharedMemory )                                           [    0xffff668d33c0] bci=[-1,2,19] rc=1 vc=139 vn=- li=2 udi=- nc=2 flg=0x8100
 n20n     (  1)        ==>aRegLoad (in &GPR_0016) (X!=0 SeenRealReference sharedMemory )
 n12n     (  1)        lconst 24 (highWordZero X!=0 X>=0 )                                            [    0xffff668d3370] bci=[-1,2,19] rc=1 vc=139 vn=- li=2 udi=- nc=0 flg=0x4104
------------------------------
------------------------------
 n15n     (  0)  ireturn                                                                              [    0xffff668d3460] bci=[-1,3,19] rc=0 vc=139 vn=- li=2 udi=- nc=1
 n14n     (  0)    iloadi  <array-shadow>[#204  Shadow] [flags 0x80000603 0x0 ] (in GPR_0018) (cannotOverflow )  [    0xffff668d3410] bci=[-1,2,19] rc=0 vc=139 vn=- li=2 udi=12432 nc=1 flg=0x1000
 n13n     (  1)      aladd (X>=0 internalPtr sharedMemory )                                           [    0xffff668d33c0] bci=[-1,2,19] rc=1 vc=139 vn=- li=2 udi=- nc=2 flg=0x8100
 n20n     (  0)        ==>aRegLoad (in &GPR_0016) (X!=0 SeenRealReference sharedMemory )
 n12n     (  0)        lconst 24 (highWordZero X!=0 X>=0 )                                            [    0xffff668d3370] bci=[-1,2,19] rc=0 vc=139 vn=- li=2 udi=- nc=0 flg=0x4104
------------------------------

 [    0xffff66963150]   2       ldrimmw         GPR_0018, [&GPR_0016, 24]               # SymRef  <array-shadow>[#204  Shadow] [flags 0x80000603 0x0 ]  
 [    0xffff66963920]   2       assocreg[&GPR_0016 : x0] 
 [    0xffff66963240]   3       retn 
 PRE: [GPR_0018 : x0] 
POST: 
```

**Comment by knn-k** [comment]
I recreated the case with `aladd` above locally.

**Comment by knn-k** [comment]
I opened #6232 for the `aladd` case above.

**Comment by knn-k** [comment]
This has been implemented by #6274.
Thank you, @Akira1Saitoh !
